File: packages/agent-core/src/agent/injection/parts-manifest.ts
/**
 * parts-manifest.ts — mode-agnostic parsing for the Parts manifest that drives
 * the multi-file split protocol.
 *
 * Both plan mode and design mode write a split INDEX whose durable state is a
 * markdown "Parts" table (one row per sibling file, each `pending`/`done`).
 * The parsing here is purely textual and carries no plan/design-specific
 * wording, so both session modes import it. The split *directives* (the prompt
 * text steering the model to the next part) stay in each mode's contract file,
 * since their wording differs (ExitPlanMode vs ExitDesignMode, "task" vs
 * "component").
 */

import { basename } from 'pathe';

export interface ManifestPart {
  readonly file: string;
  readonly scope: string;
}

export interface PartsManifest {
  /** True when a manifest table exists and every row is `done`. */
  readonly allDone: boolean;
  /** The first `pending` row, or null when none remain. */
  readonly next: ManifestPart | null;
}

export interface ManifestCounts {
  /** Number of rows marked `done`. */
  readonly done: number;
  /** Number of rows marked `pending`. */
  readonly pending: number;
}

interface ManifestRow {
  readonly file: string;
  readonly scope: string;
  readonly status: 'pending' | 'done';
}

/**
 * Scan split-index content for Parts-manifest table rows. Resilient to surrounding
 * prose: a row counts only when its last cell is `pending`/`done` and its file cell
 * ends in `.md`, so header and separator rows are ignored. The single source of truth
 * shared by every manifest accessor below.
 */
function scanManifestRows(content: string): ManifestRow[] {
  const rows: ManifestRow[] = [];
  for (const line of content.split('\n')) {
    const cells = line.split('|').map((cell) => cell.trim());
    // Drop the empty cells produced by leading/trailing pipes.
    const trimmed = cells.filter(
      (cell, index) => !(cell === '' && (index === 0 || index === cells.length - 1)),
    );
    if (trimmed.length < 4) continue;
    const status = (trimmed.at(-1) ?? '').toLowerCase();
    if (status !== 'pending' && status !== 'done') continue;
    const file = (trimmed[1] ?? '').replaceAll('`', '').trim();
    if (!file.toLowerCase().endsWith('.md')) continue;
    rows.push({ file, scope: trimmed.at(-2) ?? '', status });
  }
  return rows;
}

/**
 * Parse a Parts manifest out of split-index content.
 * Returns null when no manifest table is present (a single-file plan/design).
 */
export function parsePartsManifest(content: string): PartsManifest | null {
  const rows = scanManifestRows(content);
  if (rows.length === 0) return null;
  const next = rows.find((row) => row.status === 'pending');
  return {
    allDone: next === undefined,
    next: next === undefined ? null : { file: basename(next.file), scope: next.scope },
  };
}

/**
 * Basenames of EVERY sibling file listed in a split-index Parts manifest (not just
 * the next pending one). Used to gather a split plan/design's sibling files for review.
 * Returns `[]` when there is no manifest table (a single-file plan/design).
 */
export function parseManifestFiles(content: string): string[] {
  return scanManifestRows(content).map((row) => basename(row.file));
}

/**
 * Counts of `done`/`pending` rows in a split-index Parts manifest, or null when
 * there is no manifest table (a single-file plan/design). Used to detect part
 * boundaries (a `done` count that grew since the previous step).
 */
export function countManifestRows(content: string): ManifestCounts | null {
  const rows = scanManifestRows(content);
  if (rows.length === 0) return null;
  let done = 0;
  let pending = 0;
  for (const row of rows) {
    if (row.status === 'done') done++;
    else pending++;
  }
  return { done, pending };
}

File: rust-ody/crates/agent-rs/src/injection/parts_manifest.rs
/// A single row in the parts manifest table.
#[derive(Debug, Clone, PartialEq)]
pub struct ManifestPart {
    pub file: String,
    pub scope: String,
}

/// Parsed parts manifest — mirrors TS `PartsManifest`.
#[derive(Debug, Clone, PartialEq)]
pub struct PartsManifest {
    pub all_done: bool,
    pub next: Option<ManifestPart>,
}

/// Parse a markdown table with columns `# | File | Scope | Status` to find the first pending row.
/// Mirrors TS `parsePartsManifest`.
pub fn parse_parts_manifest(content: &str) -> Option<PartsManifest> {
    let rows: Vec<(&str, &str, &str)> = content
        .lines()
        .filter(|line| line.starts_with('|') && !line.starts_with("|---") && !line.contains("File"))
        .filter_map(|line| {
            let cells: Vec<&str> = line.split('|').map(|s| s.trim()).collect();
            // cells[0] is empty (leading |), cells[1] is #, cells[2] is File, cells[3] is Scope, cells[4] is Status
            if cells.len() >= 5 {
                let file = cells.get(2).unwrap_or(&"");
                let scope = cells.get(3).unwrap_or(&"");
                let status = cells.get(4).unwrap_or(&"");
                if !file.is_empty() {
                    Some((*file, *scope, *status))
                } else {
                    None
                }
            } else {
                None
            }
        })
        .collect();

    if rows.is_empty() {
        return None;
    }

    let all_done = !rows
        .iter()
        .any(|(_, _, status)| status.to_lowercase() == "pending");

    let next = if all_done {
        None
    } else {
        rows.iter()
            .find(|(_, _, status)| status.to_lowercase() == "pending")
            .map(|(file, scope, _)| ManifestPart {
                file: file.to_string(),
                scope: scope.to_string(),
            })
    };

    Some(PartsManifest { all_done, next })
}

/// Extract all file names from a parts manifest table.
/// Mirrors TS `parseManifestFiles`.
pub fn parse_manifest_files(content: &str) -> Vec<String> {
    content
        .lines()
        .filter(|line| line.starts_with('|') && !line.starts_with("|---") && !line.contains("File"))
        .filter_map(|line| {
            let cells: Vec<&str> = line.split('|').map(|s| s.trim()).collect();
            let file = cells.get(2).unwrap_or(&"");
            if file.is_empty() {
                None
            } else {
                Some(file.to_string())
            }
        })
        .collect()
}

/// Count done/pending rows — mirrors TS `countManifestRows`.
pub fn count_manifest_rows(content: &str) -> Option<(usize, usize)> {
    let mut done = 0usize;
    let mut pending = 0usize;
    let mut found_table = false;
    for line in content.lines() {
        if line.starts_with('|') && !line.starts_with("|---") && !line.contains("File") {
            found_table = true;
            let cells: Vec<&str> = line.split('|').map(|s| s.trim()).collect();
            match cells.get(4).map(|s| s.to_lowercase()).as_deref() {
                Some("done") => done += 1,
                Some("pending") => pending += 1,
                _ => {}
            }
        }
    }
    if found_table {
        Some((done, pending))
    } else {
        None
    }
}
